#In Class Assignment 2

In [164]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [165]:
data = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/newhousing.csv")

In [166]:
data.head()

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  basement  \
0  5250000  5500         3          2        1         1          0         1   
1  4480000  4040         3          1        2         1          0         0   
2  3570000  3640         2          1        1         1          0         0   
3  2870000  3040         2          1        1         0          0         0   
4  3570000  4500         2          1        1         0          0         0   

   hotwaterheating  airconditioning  parking  prefarea  semi-furnished  \
0                0                0        0         0               1   
1                0                0        1         0               1   
2                0                0        0         0               0   
3                0                0        0         0               0   
4                0                0        0         0               0   

   unfurnished  areaperbedroom   bbratio  
0            0     1833.333333  0.666667  
1            0     1346.666667  0.333333  
2            1     1820.000000  0.500000  
3            1     1520.000000  0.500000  
4            0     2250.000000  0.500000

In [167]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   price            545 non-null    int64  
 1   area             545 non-null    int64  
 2   bedrooms         545 non-null    int64  
 3   bathrooms        545 non-null    int64  
 4   stories          545 non-null    int64  
 5   mainroad         545 non-null    int64  
 6   guestroom        545 non-null    int64  
 7   basement         545 non-null    int64  
 8   hotwaterheating  545 non-null    int64  
 9   airconditioning  545 non-null    int64  
 10  parking          545 non-null    int64  
 11  prefarea         545 non-null    int64  
 12  semi-furnished   545 non-null    int64  
 13  unfurnished      545 non-null    int64  
 14  areaperbedroom   545 non-null    float64
 15  bbratio          545 non-null    float64
dtypes: float64(2), int64(14)
memory usage: 68.2 KB


In [168]:
data.columns

Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'semi-furnished', 'unfurnished',
       'areaperbedroom', 'bbratio'],
      dtype='object')

## Splitting the Data

In [169]:
X = data.drop('price', axis = 1)
y = data[['price']]

##Train Test Split

In [170]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, 
                                                    random_state = 0)

## Feature Scaling

In [171]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_scaled = sc.fit_transform(X)

In [172]:
from sklearn.decomposition import PCA
pca = PCA(n_components= 2)
pca.fit(X_scaled)
x_pca = pca.transform(X_scaled)

## Linear Regression

In [173]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

In [174]:
y_pred = regressor.predict(X_test)

In [175]:
regressor.score(X_train, y_train)

0.7207292197818493

In [176]:
regressor.score(X_test, y_test)

0.5481622196535569

## Random Forest Regression

In [177]:
from sklearn.ensemble import RandomForestRegressor
RFG = RandomForestRegressor()
RFG.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


RandomForestRegressor()

In [178]:
RFG.score(X_train, y_train)

0.9493403902064194

In [179]:
RFG.score(X_test, y_test)

0.528963476560675

## PCA Train Test Split

In [180]:
X_train, X_test, y_train, y_test = train_test_split(x_pca, y, test_size = 0.3,
                                                    random_state = 0)

## PCA with Linear Regression

In [181]:
regressor.fit(X_train, y_train)

LinearRegression()

In [182]:
regressor.score(X_train, y_train)

0.6717632678848646

In [185]:
regressor.score(X_test, y_test)

0.5699901167619912

## PCA with Random Forest Regression

In [183]:
RFG.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.


RandomForestRegressor()

In [184]:
RFG.score(X_train, y_train)

0.9420518883799532

In [186]:
RFG.score(X_test, y_test)

0.475063532883774

### For the experiment I found Random Forest to be the best model for housing price calculation with a score of 95% though the model may be overfitted. In that case the best model would be Linear Regression with a score of 72%